### ISSUE TITLE
Performance when guessing delimiter with quotes

### ISSUE THREAD

**Issue by derek-bo-derek** [mholt/PapaParse#624]
I'm seeing a performance problem when parsing a moderately sized file without a declared delimiter.

I found that PapaParse tries parsing the file with a handful of field delimiters and picking the most likely candidate. In the `Parser.parse` method, if it finds a quote character, it tries to match it to the end quote, but an end quote either must be right before the delimiter or before the new line for it to accept it as the end quote. 

In my case, the only field that is quoted is the first column in the file:
```
"timestamp","field1","field2","field3"
"2018-02-21 19:20:00",1.1,2.2,3.3
"2018-02-21 19:30:00",4.4,5.5,6.6
"2018-02-21 19:30:00",7.7,8.8,9.9
```

So, when it tries to parse the file with a tab delimiter, it parses the first row with a single field by matching the first quote with the quote on the end of the line. Then it goes to the second row, and never finds an end quote, because every quote in the rest of the file doesn't end before either a tab or an end of line. 

The performance isn't noticable on https://www.papaparse.com/demo (at least not for a 19.5MB files). On the application I'm working on, it freezes the tab for a very long time due to a polyfill for trim methods on Strings. I've been able to fix the problem for myself by setting an explicit comma delimiter for the files I'm parsing.

Would PapaParse be better off if it matches quotes more efficiently? At least if it would stop trying to match the quote if there is a new line delimiter?

**Comment by pokoli** [comment]
Hi @derek-bo-derek, thank you for oppening this issue. 

Yes, we will be happy to acccept a patch that stops searching for the quote char if a new line is found. 
Please make sure to include a test to ensure the behaviour stays for future versions.

**Comment by derek-bo-derek** [comment]
I looked into making the changes, but there are tests that accept new lines within quotes (eg. "Quoted field with line break"), which made me realize that makes sense depending on the data in the CSV (eg. multiline text fields).

I suppose that handling quoted fields differently is a tricky when you don't have the correct delimiter. It looks like the existing code tries to assume that a quote that isn't immediately before a delimiter or new line is just malformed and that there is a valid end quote later in the file.

Perhaps the "guess delimiter" functionality could be separated from the parsing functionality? Hard for me to say what would be a better heuristic for picking the best delimiter.

I'm going to leave this alone for now since I understand more about why the parser is doing what it is doing.

**Comment by dboskovic** [comment]
This is the same thing that cropped up in #562 - that polyfill is awful.

**Comment by FloPrey** [comment]
Is there a possible fix for this? We have the same behavior in our application but without using babel-polyfill. I have the same issue that it's freezing because of the delimiter guessing new line issue.
The only solution I have now is to downgrade to version 4.3.

**Comment by janisdd** [comment]
I recently ran into the same issue... I think it's the same or at a least related issue.
An easy way to reproduce this is to make a relatively large file (several MB), use | as delimiter and quote all fields and use CRLF as new line character.
Then remove the closing quotes from the last field in the first row, e.g.

```
"xx"|"zz
"yy"|"zz"
...
```

Some background:
The first think to note is that the new line character is guessed wrong (LF instead of CRLF). This is because inside the guessLineEndings code everything between "..." is removed. Because of the missing quote every newline characters are removed -> defaults to LF. (I'm also not 100% what `([^]*?)` is actually matching as it is not a valid regex but somehow it works).

Inside guessDelimiter we use a preview of 10 rows.
But because of the wrong new line character the search for the closing quote is never finished (only when at EOF). The loop never ends because `// Closing quote followed by newline ... condition` is never met (we expect LF as new line character but actual is CRLF).
The loop then examines all quotes in the whole file for every delimiter.


I fixed this in my version by adding an option to the Parser config `isGuessingDelimiter: bool` and `maxGuessLength: number`.
`isGuessingDelimiter` is only set when we guess the delimiter and `maxGuessLength` is a limit for `quoteSearch`.

TL;DR
Specifically I would propose these changes (based on version 5747da6 on 25 Aug):

at Line: 1277:

```diff
var preview = new Parser({
	comments: comments,
	delimiter: delim,
	newline: newline,
-	preview: 10
+	preview: 10,
+	isGuessingDelimiter: true
}).parse(input);
```

after Line: 1378:
```js
if (typeof config.isGuessingDelimiter === 'boolean') {
	isGuessingDelimiter = config.isGuessingDelimiter;

	if (typeof config.maxGuessLength === 'number') {
		maxGuessLength = config.maxGuessLength;
	} else {
		//default to 5000?
		maxGuessLength = 5000;
	}
}
```

at Line: 1464:

```js
//if the text does not contain the delimiter (not even in quoted fields) we can return early
if (isGuessingDelimiter && nextDelim === -1 && cursor === 0) {
	return finish('');
}
```

at Line 1486:

```js
// we exceeded the max search length for the delimiter, give up
if (isGuessingDelimiter && maxGuessLength && quoteSearch > maxGuessLength) {
	return finish('');
}
```
This will terminate the quote search after position 5000 when guessing the delimiter.
